waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/StaticDatabaseMappingService.java

```diff
@@ -94,54 +95,53 @@ public StaticDatabaseMappingService(
     init(initialMetastores);
   }
 
-  private void init(List<AbstractMetaStore> federatedMetaStores) {
+  private void init(List<AbstractMetaStore> metaStores) {
     mappingsByMetaStoreName = Collections.synchronizedMap(new LinkedHashMap<>());
-    mappingsByDatabaseName = new ConcurrentHashMap<>();
-    for (AbstractMetaStore federatedMetaStore : federatedMetaStores) {
+    mappingsByDatabaseName = Collections.synchronizedMap(new LinkedHashMap<>());
+    databaseMappingToDatabaseList = new ConcurrentHashMap<>();
+    for (AbstractMetaStore federatedMetaStore : metaStores) {
       add(federatedMetaStore);
     }
   }
 
   private void add(AbstractMetaStore metaStore) {
     MetaStoreMapping metaStoreMapping = metaStoreMappingFactory.newInstance(metaStore);
 
+    List<String> mappableDatabases = Collections.emptyList();
+    if (metaStoreMapping.isAvailable()) {
+      try {
+        List<String> allDatabases = metaStoreMapping.getClient().get_all_databases();
+        Whitelist whitelistedDatabases = new Whitelist(metaStore.getMappedDatabases());
+        mappableDatabases = applyWhitelist(allDatabases, whitelistedDatabases);
+      } catch (TException e) {
+        LOG.error("Could not get databases for metastore {}", metaStore.getRemoteMetaStoreUris(), e);
+      }
+    }
+    DatabaseMapping databaseMapping = createDatabaseMapping(metaStoreMapping);
+
     if (metaStore.getFederationType() == PRIMARY) {
-      validatePrimaryMetastoreDatabases(metaStoreMapping);
-      primaryDatabaseMapping = createDatabaseMapping(metaStoreMapping);
+      validatePrimaryMetastoreDatabases(mappableDatabases);
+      primaryDatabaseMapping = databaseMapping;
       primaryDatabasesCache.invalidateAll();
-      mappingsByMetaStoreName.put(metaStoreMapping.getMetastoreMappingName(), primaryDatabaseMapping);
     } else {
-      FederatedMetaStore federatedMetaStore = (FederatedMetaStore) metaStore;
-      List<String> mappableDatabases = Collections.emptyList();
-      if (metaStoreMapping.isAvailable()) {
-        try {
-          List<String> allFederatedDatabases = metaStoreMapping.getClient().get_all_databases();
-          mappableDatabases = applyWhitelist(allFederatedDatabases, federatedMetaStore.getMappedDatabases());
-        } catch (TException e) {
-          LOG.error("Could not get databases for metastore {}", federatedMetaStore.getRemoteMetaStoreUris(), e);
-        }
-      }
       validateFederatedMetastoreDatabases(mappableDatabases, metaStoreMapping);
-      DatabaseMapping databaseMapping = createDatabaseMapping(metaStoreMapping);
-      mappingsByMetaStoreName.put(metaStoreMapping.getMetastoreMappingName(), databaseMapping);
-      addDatabaseMappings(mappableDatabases, databaseMapping);
     }
+
+    mappingsByMetaStoreName.put(metaStoreMapping.getMetastoreMappingName(), databaseMapping);
+    addDatabaseMappings(mappableDatabases, databaseMapping);
+    databaseMappingToDatabaseList.put(databaseMapping, mappableDatabases);
```

Comment:
I think we should do this put on the key: `databaseMapping.getMetastoreMappingName()`. So we don't rely on the object instance equals of a DataMapping object.

Reply:
Good point, thanks :)